lib/min_heap.rb

```diff
@@ -35,37 +45,106 @@ def to_s
 
     output = "["
     (@store.length - 1).times do |index|
-      output += @store[index].value + ", "
+      output += @store[index].value.to_s + ", "
     end
 
-    output += @store.last.value + "]"
+    output += @store.last.value.to_s + "]"
       
     return output
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(n)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.length == 0
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(1)
   def heap_up(index)
+    if index == 0
+      # no more upstream to compare to
+      return
     
+    else
+      # Compare node @ index to its parent
+      # if child >= parent, min heap :-)  base case
+      # if child < parent, swap places, then heap_up(parentIndex), recursion!
+
+      # this gets the correct ParentIndex for either L or R childIndex
+        # ex: for parent @ index 1, its LC should be at index 3 and RC at index 4
+          # calculating the parentIndex for index 3... we get (0.5).ceil = 1
+          # and calculating the parentIndex for index 4... we get 1.ceil = 1
+      parentIndex = ((index - 2).to_f / 2).ceil
+
+      if @store[index].key >= @store[parentIndex].key
+        return 
+      else
+        swap(index, parentIndex)
+        heap_up(parentIndex)
+        return
+      end
+    end
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
-  def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+  # Time complexity: O(log n)
+  # Space complexity: O(1)
+  def heap_down(parentIndex)
+    # compare node @ parentIndex to its children
+    # if parent <= both children, min heap :-)  base case
+    # if parent > either/both child, swap places with the smaller child, then min_heap_down(childIndex), recursion!
+    # when no more children to compare to, base case :-)
+
+    # first find out if parent has LC or RC
+    indexLC = parentIndex * 2 + 1 
+    indexRC = indexLC + 1
```

Comment:
Just note for ruby these should be named:  `index_lc` and `parent_index` and `index_rc` etc.